File: Example/javaapp/src/main/java/com/stripe/example/javaapp/fragment/discovery/ReaderHolder.java
package com.stripe.example.javaapp.fragment.discovery;

import android.content.res.Resources;
import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

import com.stripe.example.javaapp.R;
import com.stripe.stripeterminal.external.models.Location;
import com.stripe.stripeterminal.external.models.Reader;
import com.stripe.example.javaapp.databinding.ListItemCardBinding;

import org.jetbrains.annotations.NotNull;

/**
 * A simple [RecyclerView.ViewHolder] that also acts as a [View.OnClickListener] to allow for
 * selecting a reader.
 */
public class ReaderHolder extends RecyclerView.ViewHolder {
    @NotNull private final ReaderClickListener clickListener;
    @NotNull private final ListItemCardBinding binding;
    @NotNull private final Resources resources;

    public ReaderHolder(
        @NotNull View parent,
        @NotNull ReaderClickListener clickListener
    ) {
        super(parent);
        this.binding = ListItemCardBinding.bind(parent);
        this.resources = parent.getResources();
        this.clickListener = clickListener;
    }

    void bind(@NotNull Reader reader, Location location) {
        binding.listItemCardTitle.setText(
            reader.getSerialNumber() != null ? reader.getSerialNumber() : reader.getId()
        );
        if (location == null && reader.getLocation() == null) {
            binding.listItemCardDescription.setText(R.string.discovery_reader_location_unavailable);
        } else if (location == null) {
            binding.listItemCardDescription.setText(
                resources.getString(
                    R.string.discovery_reader_location_last,
                    reader.getLocation().getDisplayName()
                )
            );
        } else {
            binding.listItemCardDescription.setText(
                resources.getString(R.string.discovery_reader_location, location.getDisplayName())
            );
        }
        binding.listItemCard.setOnClickListener((View.OnClickListener) v -> clickListener.onClick(reader));
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/fragment/discovery/ReaderHolder.kt
package com.stripe.example.fragment.discovery

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.stripe.example.R
import com.stripe.example.databinding.ListItemCardBinding
import com.stripe.stripeterminal.external.models.Location
import com.stripe.stripeterminal.external.models.Reader

/**
 * A simple [RecyclerView.ViewHolder] that also acts as a [View.OnClickListener] to allow for
 * selecting a reader.
 */
class ReaderHolder(
    parent: View,
    private val clickListener: ReaderClickListener,
) : RecyclerView.ViewHolder(parent) {
    private val binding = ListItemCardBinding.bind(parent)
    private val resources = parent.resources

    fun bind(reader: Reader, locationSelection: Location?) {
        binding.listItemCardTitle.text = reader.serialNumber
            ?: reader.id
            ?: resources.getString(R.string.discovery_reader_unknown)
        binding.listItemCardDescription.text = when {
            locationSelection == null && reader.location == null -> resources.getString(
                R.string.discovery_reader_location_unavailable
            )
            locationSelection == null -> resources.getString(
                R.string.discovery_reader_location_last,
                reader.location!!.displayName
            )
            else -> resources.getString(
                R.string.discovery_reader_location,
                locationSelection.displayName
            )
        }
        binding.listItemCard.setOnClickListener {
            clickListener.onClick(reader)
        }
    }
}
